File: crates/fc-sdk/src/client/dispatch_pools.rs
//! Dispatch pool management operations.

use serde::{Deserialize, Serialize};
use super::{FlowCatalystClient, ClientError, ListResponse};

/// Request to create a dispatch pool.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct CreateDispatchPoolRequest {
    pub code: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub client_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rate_limit: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub concurrency: Option<u32>,
}

/// Request to update a dispatch pool.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UpdateDispatchPoolRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rate_limit: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub concurrency: Option<u32>,
}

/// Filters for listing dispatch pools.
#[derive(Debug, Clone, Default)]
pub struct DispatchPoolFilters {
    pub client_id: Option<String>,
    pub status: Option<String>,
}

/// Dispatch pool response from the platform API.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DispatchPoolResponse {
    pub id: String,
    pub code: String,
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub client_id: Option<String>,
    pub status: String,
    #[serde(default)]
    pub rate_limit: Option<u32>,
    #[serde(default)]
    pub concurrency: Option<u32>,
    pub created_at: String,
    pub updated_at: String,
}

impl FlowCatalystClient {
    /// List dispatch pools with optional filters.
    pub async fn list_dispatch_pools(
        &self,
        filters: &DispatchPoolFilters,
    ) -> Result<ListResponse<DispatchPoolResponse>, ClientError> {
        let mut params = Vec::new();
        if let Some(ref cid) = filters.client_id {
            params.push(format!("clientId={}", cid));
        }
        if let Some(ref s) = filters.status {
            params.push(format!("status={}", s));
        }
        let query = if params.is_empty() {
            String::new()
        } else {
            format!("?{}", params.join("&"))
        };
        self.get(&format!("/api/dispatch-pools{}", query))
            .await
    }

    /// Get a dispatch pool by ID.
    pub async fn get_dispatch_pool(
        &self,
        id: &str,
    ) -> Result<DispatchPoolResponse, ClientError> {
        self.get(&format!("/api/dispatch-pools/{}", id)).await
    }

    /// Create a new dispatch pool.
    pub async fn create_dispatch_pool(
        &self,
        req: &CreateDispatchPoolRequest,
    ) -> Result<DispatchPoolResponse, ClientError> {
        self.post("/api/dispatch-pools", req).await
    }

    /// Update a dispatch pool.
    pub async fn update_dispatch_pool(
        &self,
        id: &str,
        req: &UpdateDispatchPoolRequest,
    ) -> Result<DispatchPoolResponse, ClientError> {
        self.put(&format!("/api/dispatch-pools/{}", id), req)
            .await
    }

    /// Delete a dispatch pool.
    pub async fn delete_dispatch_pool(&self, id: &str) -> Result<(), ClientError> {
        self.delete_req(&format!("/api/dispatch-pools/{}", id))
            .await
    }

    /// Suspend a dispatch pool.
    pub async fn suspend_dispatch_pool(
        &self,
        id: &str,
    ) -> Result<DispatchPoolResponse, ClientError> {
        self.post_action(&format!("/api/dispatch-pools/{}/suspend", id))
            .await
    }

    /// Activate a dispatch pool.
    pub async fn activate_dispatch_pool(
        &self,
        id: &str,
    ) -> Result<DispatchPoolResponse, ClientError> {
        self.post_action(&format!("/api/dispatch-pools/{}/activate", id))
            .await
    }
}

File: frontend/src/api/dispatch-pools.ts
import { apiFetch } from "./client";
import type {
	CreatedResponse,
	DispatchPoolListResponse as GenDispatchPoolListResponse,
	DispatchPoolResponse,
} from "./generated";

// Request-side string union the forms/filters rely on. The generated
// response types deliberately stay `string` (the spec doesn't carry enums —
// see docs/frontend-api-types-adoption.md on SDK coordination).
export type DispatchPoolStatus = "ACTIVE" | "SUSPENDED" | "ARCHIVED";

// Response types alias the generated contract (api/openapi.lock.json) so
// `vue-tsc` fails on backend drift. Aliased under the historical names so
// pages keep their imports. `rateLimit` is absent (not `null`) on the wire
// for concurrency-only pools.
export type DispatchPool = DispatchPoolResponse;
export type DispatchPoolListResponse = GenDispatchPoolListResponse;

export interface CreateDispatchPoolRequest {
	code: string;
	name: string;
	description?: string;
	/** Optional. Omit/undefined for concurrency-only pools. */
	rateLimit?: number;
	concurrency: number;
	clientId?: string;
}

export interface UpdateDispatchPoolRequest {
	name?: string;
	description?: string;
	rateLimit?: number;
	concurrency?: number;
	status?: DispatchPoolStatus;
}

export interface DispatchPoolFilters {
	clientId?: string;
	status?: DispatchPoolStatus;
	anchorLevel?: boolean;
}

export const dispatchPoolsApi = {
	list(filters: DispatchPoolFilters = {}): Promise<DispatchPoolListResponse> {
		const params = new URLSearchParams();
		if (filters.clientId) params.set("clientId", filters.clientId);
		if (filters.status) params.set("status", filters.status);
		if (filters.anchorLevel !== undefined)
			params.set("anchorLevel", String(filters.anchorLevel));

		const query = params.toString();
		return apiFetch(`/dispatch-pools${query ? `?${query}` : ""}`);
	},

	get(id: string): Promise<DispatchPool> {
		return apiFetch(`/dispatch-pools/${id}`);
	},

	/** POST /dispatch-pools returns the standard created envelope `{ id }`, not the full pool. */
	create(data: CreateDispatchPoolRequest): Promise<CreatedResponse> {
		return apiFetch("/dispatch-pools", {
			method: "POST",
			body: JSON.stringify(data),
		});
	},

	update(id: string, data: UpdateDispatchPoolRequest): Promise<void> {
		return apiFetch(`/dispatch-pools/${id}`, {
			method: "PUT",
			body: JSON.stringify(data),
		});
	},

	// delete/suspend/activate return 204 No Content on the wire; the old
	// `{ message, poolId }` envelope never existed.
	delete(id: string): Promise<void> {
		return apiFetch(`/dispatch-pools/${id}`, {
			method: "DELETE",
		});
	},

	suspend(id: string): Promise<void> {
		return apiFetch(`/dispatch-pools/${id}/suspend`, {
			method: "POST",
		});
	},

	activate(id: string): Promise<void> {
		return apiFetch(`/dispatch-pools/${id}/activate`, {
			method: "POST",
		});
	},
};
